### PR TITLE
[BEAM-5865] Create optional auto-balancing sharding function for Flink

### PR DESCRIPTION
**Pull request by JozoVilcek**
This change has to parts.
* modify WriteFiles to be able to specify sharding function
* introduce an option for FlinkRunner to use sharding function which distributes shards evenly among workers

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by JozoVilcek** [comment]
retest this please

**Comment by JozoVilcek** [comment]
@reuvenlax @mxm

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 1]
```diff
@@ -211,6 +254,23 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
...
           replacement = replacement.withWindowedWrites();
         }
+
+        if (options.isAutoBalanceWriteFilesShardingEnabled()) {
+          Preconditions.checkArgument(
+              options.getParallelism() > 0,
+              "Parallelism is required to be set in FlinkPipelineOptions when isAutoBalanceWriteFilesShardingEnabled");
+          Preconditions.checkArgument(
+              options.getMaxParallelism() > 0,
+              "MaxParallelism is required to be set in FlinkPipelineOptions when isAutoBalanceWriteFilesShardingEnabled");
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Why is this required, couldn't we assume `maxParallelism == parallelism` if not provided?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 2]
```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
...
     public Map<PValue, ReplacementOutput> mapOutputs(
         Map<TupleTag<?>, PValue> outputs, WriteFilesResult<DestinationT> newOutput) {
-      return Collections.emptyMap();
+      return ReplacementOutputs.tagged(outputs, newOutput);
+    }
+  }
+
+  /**
+   * Flink has a known problem of unevenly assigning keys to key groups (and then workers) for cases
+   * that number of keys is not >> key groups. This is typical scenario when writing files, where
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: I wouldn't say this is a Flink problem, it just comes with the nature of hashing algorithms (Murmur hash in this case) when the number of keys is not much bigger than the number of partitions.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 3]
```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
...
+      if (cache.size() < shardNumber) {
+        for (int i = 0; i < shardNumber; i++) {
+          generateInternal(new CacheKey(destinationKey, i));
+        }
+      }
+
+      return generateInternal(new CacheKey(destinationKey, shardNumber));
+    }
+
+    private ShardedKey<Integer> generateInternal(CacheKey key) {
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Rename to `getShardedKey` and add a Precondition that the cache has been created upfront.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 4]
```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
...
+
+      int destinationKey =
+          Arrays.hashCode(CoderUtils.encodeToByteArray(destinationCoder, destination));
+      // we need to ensure that keys are always stable no matter at which worker they
+      // are created and what is an order of observed shard numbers
+      if (cache.size() < shardNumber) {
+        for (int i = 0; i < shardNumber; i++) {
+          generateInternal(new CacheKey(destinationKey, i));
+        }
+      }
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Could be refactored to a method named `generateInitialShardedKeys` which only takes care of the initial generation.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 5]
```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
...
+      while (true) {
+        salt++;
+        ShardedKey<Integer> shk = ShardedKey.of(Objects.hash(key.key, salt), key.shard);
+        int targetPartition = key.shard % parallelism;
+
+        // create effective key in the same way Beam/Flink will do so we can see if it gets
+        // allocated to the partition we want
+        ByteBuffer effectiveKey;
+        try {
+          byte[] bytes = CoderUtils.encodeToByteArray(ShardedKeyCoder.of(VarIntCoder.of()), shk);
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: use `FlinkKeyUtils`

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 6]
```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
...
+   * workers by Flink.
+   */
+  private static class FlinkAutoBalancedShardKeyShardingFunction<UserT, DestinationT>
+      implements ShardingFunction<UserT, DestinationT> {
+
+    private final int parallelism;
+    private final int maxParallelism;
+    private final Coder<DestinationT> destinationCoder;
+    private final Map<CacheKey, ShardedKey<Integer>> cache = new HashMap<>();
+    private final Set<Integer> usedSalts = new HashSet<>();
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Does this still serve a purpose?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 7]
```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
...
+      this.maxParallelism = maxParallelism;
+      this.destinationCoder = destinationCoder;
+    }
+
+    @Override
+    public ShardedKey<Integer> assignShardKey(
+        DestinationT destination, UserT element, int shardCount) throws Exception {
+
+      if (shardNumber == -1) {
+        shardNumber = ThreadLocalRandom.current().nextInt(shardCount);
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Is there any benefit to starting with a random shard number as opposed to `0`?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java [hunk 8]
```diff
@@ -278,6 +280,36 @@ public void testExpandShardedWrite() throws IOException {
         WriteFiles.to(makeSimpleSink()).withNumShards(20));
   }
 
+  /**
+   * Test that WriteFiles with a configured sharding function distributes elements according to
+   * assigned key. Test use simple sharding which co-locate same elements into same shard.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testCustomShardingFunctionWrite() throws IOException {
```
Comment (mxm) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java]: Could we also add a test for `FlinkAutoBalancedShardKeyShardingFunction`?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 9]
```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
...
+          effectiveKey = ByteBuffer.wrap(bytes);
+        } catch (CoderException e) {
+          throw new RuntimeException(e);
+        }
+
+        int partition =
+            KeyGroupRangeAssignment.assignKeyToParallelOperator(
+                effectiveKey, maxParallelism, parallelism);
+
+        if (partition == targetPartition && !usedSalts.contains(salt)) {
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: I know this will succeed by likelihood but I'm wondering whether we should put an upper limit here in case the loop does not return.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java (hunk 1 above)
Comment (JozoVilcek) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: I think Flik does use `KeyGroupRangeAssignment .computeDefaultMaxParallelism()` and not directly parallelism. I guess I was just being conservative. I will make it work with defaults

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java (hunk 2 above)
Comment (JozoVilcek) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: True. Does this sounds better?
```
Flink assigns elements to parallel operators (workers) based their key group. Key group is determined based on a Murmur hash of element's key. This will have skew distribution in case when number of keys is not >> key groups, which is typical scenario when writing files, where one do not want to end up with too many small files.
```

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java (hunk 7 above)
Comment (JozoVilcek) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: I just took it from WriteFiles, where it says
```
// We want to desynchronize the first record sharding key for each instance of
// ApplyShardingKey, so records in a small PCollection will be statistically balanced.
```

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java (hunk 8 above)
Comment (JozoVilcek) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java]: Yes

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java (hunk 1 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Ah that's true. I wasn't aware that we set a higher max parallelism, just in case the user wants to scale up.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java (hunk 2 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Sounds good!

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 10]
```diff
@@ -325,74 +357,63 @@ private FlinkAutoBalancedShardKeyShardingFunction(
...
+          }
+          ShardedKey<Integer> shk = ShardedKey.of(Objects.hash(key, salt), shard);
+          int targetPartition = shard % parallelism;
+
+          // create effective key in the same way Beam/Flink will do so we can see if it gets
+          // allocated to the partition we want
+          ByteBuffer effectiveKey;
+          try {
+            byte[] bytes = CoderUtils.encodeToByteArray(ShardedKeyCoder.of(VarIntCoder.of()), shk);
+            effectiveKey = ByteBuffer.wrap(bytes);
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Could you use `FlinkKeyUtils.encodeKey(K key, Coder<K> keyCoder)` here? Just to ensure that this work correctly in case we change the encoding.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 11]
```diff
@@ -325,74 +357,63 @@ private FlinkAutoBalancedShardKeyShardingFunction(
...
+          } catch (CoderException e) {
+            throw new RuntimeException(e);
+          }
+
+          int partition =
+              KeyGroupRangeAssignment.assignKeyToParallelOperator(
+                  effectiveKey, maxParallelism, parallelism);
+
+          if (partition == targetPartition && !usedSalts.contains(salt)) {
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Why do we need `!usedSalts.contains(salt)`?

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+/** Tests if overrides are properly applied. */
+public class FlinkStreamingPipelineTranslatorTest {
+
+  @Test
+  public void testAutoBalanceShardKeyResolvesMaxParallelism() {
+
+    assertThat(
+        new FlinkAutoBalancedShardKeyShardingFunction<>(3, -1, StringUtf8Coder.of())
+            .getMaxParallelism(),
+        equalTo(128));
```
Comment (mxm) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java]: `equalTo(FlinkKeyGroupAssignment.getMaxParallelism())`?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java (hunk 11 above)
Comment (JozoVilcek) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: shard key identity is defined by destination (it's hash), shardNumber and salt. Salt is integer from we are searching for, to make sure the "real key" (destination, shardNumber) is landed to correct worker. Salt is restarted for each shard key to zero. 

I guess I was paranoid in the beginning to avoid key collisions. But right now, I do not see a reason why not to drop it.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java (hunk 10 above)
Comment (JozoVilcek) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Sorry, I forgot about this. Will do

Comment (JozoVilcek) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: `FlinkKeyUtils` is package private. I assume it is OK to make it public?

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java (hunk 12 above)
Comment (JozoVilcek) [reply runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java]: I can not find `` ... Can only do `equalTo(KeyGroupRangeAssignment.computeDefaultMaxParallelism(parallelism))` what is done in sharding function, if you like it better than hardcoded value.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 13]
```diff
@@ -211,6 +261,20 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
...
+              replacement.withShardingFunction(
+                  new FlinkAutoBalancedShardKeyShardingFunction<>(
+                      options.getParallelism(),
+                      options.getMaxParallelism(),
+                      sink.getDynamicDestinations().getDestinationCoder()));
+        }
+
         return PTransformReplacement.of(
             PTransformReplacements.getSingletonMainInput(transform),
             replacement.withNumShards(numShards));
```
Comment (JozoVilcek) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: I have a problem here. I found out that now, my explicit number of shards gets always override here to runner determined. 
I see that `replacement` is recreated from scratch and and must be created full. It seems that `num shards` is not explicit in payload but gets injected here and there during the transform `expand` (just guessing here)?.
It seems I can not just put sharding function to existing WriteFiles context, but needs to create fresh new one. Am I right?

I need a help here. Is there an easy way of accessing `numShards`? Do I need to change `WriteFilesPayload` and make it explicit ? @reuvenlax what approach would you suggest to resolve this?

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java (hunk 12 above)
Comment (mxm) [reply runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java]: Yes, that's what I meant.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java (hunk 13 above)
Comment (JozoVilcek) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Never mind. I talked with @mxm and noticed that overrides have access to proto and WriteFiles transform. I got derailed by runner determined sharding code and thought that accessing stuff from proto is the only way. But it seems it is now so I should be good here

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 14]
```diff
@@ -235,4 +235,11 @@
   Boolean getAllowNonRestoredState();
 
   void setAllowNonRestoredState(Boolean allowNonRestoredState);
+
+  @Description(
+      "Flag indicating whether auto-balance sharding for WriteFiles transform should be enabled.")
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Could you expand this comment to explain why this is useful and when?

**Comment by mxm** [comment]
@reuvenlax Any comments on this? I think this is good to merge otherwise.

**Comment by mxm** [comment]
@JozoVilcek Apologies, but could you rebase once more?

**Comment by mxm** [comment]
Thanks! If you want this to be included in the release, you will have to backport this to release-2.13.0.

**Comment by JozoVilcek** [comment]
Next release is fine. I can do patched versions privately. Thank for the merge and support!
